### PR TITLE
switch to Scrypt + fix error types

### PR DESCRIPTION
**Pull request by Rebal67**
issue 35

### CODE REVIEW

File: src/user.ts [hunk 1]
```diff
@@ -3,7 +3,7 @@ import { IDContract } from './services/contracts/IDContract';
 import { Bytes, KeyType, Name, PrivateKey } from '@greymass/eosio';
 import { randomBytes, randomString, sha256 } from './util/crypto';
 import { createSigner } from './services/eosio/transaction';
-import argon2 from 'argon2';
+import scrypt from "scrypt-js";
```
Comment (theblockstalk) [inline src/user.ts:6]: https://medium.com/analytics-vidhya/password-hashing-pbkdf2-scrypt-bcrypt-and-argon2-e25aaf41598e
https://crypto.stackexchange.com/a/51523

Check out these, I'd prefer to stack with argon2 if possible and even better to upgrade to argon2id as suggested

Can we try with this, an isomorphic (cross-environment) version, with argon2id?
https://www.npmjs.com/package/@sphereon/isomorphic-argon2

File: src/util/crypto.ts [hunk 2]
```diff
@@ -1,10 +1,9 @@
 import { Checksum256 } from '@greymass/eosio';
-import crypto from 'crypto';
+import * as rb from "randombytes";
```
Comment (theblockstalk) [inline src/util/crypto.ts:2]: import { default } from 'randombytes'?

File: src/user.ts (hunk 1 above)
Comment (Rebal67) [reply src/user.ts:6]: i have tested the package and just like all the other it uses pollyfill and config expo config https://github.com/react-native-community/cli/issues/870

unfortunately wasm isn't available for RN and there isn't any package that can implement that in pure javascript without crypto package

File: src/util/crypto.ts (hunk 2 above)
Comment (Rebal67) [reply src/util/crypto.ts:2]: default is a reserved word

File: package.json [hunk 3]
```diff
@@ -53,7 +54,9 @@
   },
   "dependencies": {
     "@greymass/eosio": "^0.6.2",
-    "argon2": "^0.28.7",
-    "cross-fetch": "^3.1.5"
+    "argon2": "^0.29.1",
+    "cross-fetch": "^3.1.5",
+    "randombytes": "^2.1.0",
+    "scrypt-js": "^3.0.1"
```
Comment (theblockstalk) [inline package.json:60]: remove `scrypt-js`

File: src/keymanager.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+     * generates a private key from a password and creates random salt
+     * @param password password to encrypt the private key with
+     * @returns encrypted private key and salt
+     * 
+     */
+    generatePrivateKeyFromPassword(password: string): Promise<{ privateKey: PrivateKey, salt: Buffer }>;
+
+}
+
+type KeyStorage = {
```
Comment (theblockstalk) [inline src/keymanager.ts:114]: this is specific to the implementation I think, move to the integration repo

File: src/user.ts [hunk 5]
```diff
@@ -51,19 +53,14 @@ class User {
...
-            salt
-        }
-    }
 
+    async savePassword(masterPassword: string) {
+        const { privateKey, salt } = await this.keyManager.generatePrivateKeyFromPassword(masterPassword);
+        this.salt = salt;
+        const level = KeyManagerLevel.PASSWORD;
+        this.keyManager.storeKey({ level, privateKey, challenge: masterPassword });
+    }
```
Comment (theblockstalk) [inline src/user.ts:63]: add space after this line

File: src/services/contracts/IDContract.ts [hunk 6]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (theblockstalk) [inline src/services/contracts/IDContract.ts:1]: remove line

File: src/keymanager.ts (hunk 4 above)
Comment (Rebal67) [reply src/keymanager.ts]: moved to integration

**Pull request merged**